### PR TITLE
Events Watcher API 2.0

### LINKED ISSUE #225: [SPEC] Split the `EventWatcher` trait in two traits `EventWatcher` and `EventHandler`
## Overview

The current [`EventWatcher`](https://github.com/webb-tools/relayer/blob/3eaf5e26d8ac69c7d21d125311d8cf517dda1652/src/events_watcher/mod.rs#L84) does two things:

* [`run`](https://github.com/webb-tools/relayer/blob/3eaf5e26d8ac69c7d21d125311d8cf517dda1652/src/events_watcher/mod.rs#L109): which is the main point of that trait, that runs indefinitely in the background and pulls new events for the configured contract, and this function is _implemented by default_ and the user of this trait does not need to implement this method.
* [`handle_event`](https://github.com/webb-tools/relayer/blob/3eaf5e26d8ac69c7d21d125311d8cf517dda1652/src/events_watcher/mod.rs#L92): Which is the method that needs to be implemented by the Event Watcher to react/handle that event.

This was ideal and works good, However, for only one contract and one event watcher, if we need to watch the events for the same contract but do something else, we need to construct another or a separate event watcher which means we will pull the events twice from the network.

## Goal

- The goal is to watch for an event at most ONCE and be able to handle it MULTIPLE times, instead of watching for the same event MULTIPLE times and handling it MULTIPLE times.
- The goal is to be able to implement multiple handlers for an event watcher and keep it separated.

In a simpler way to describe this, is that we need to have only one `EventWatcher` per contract and many `EventHandler` that are injected in that Event Watcher. Here is how it should look like:

```rust
/// A trait for watching events from a watchable contract.
/// EventWatcher trait exists for deployments that are smart-contract / EVM based
#[async_trait::async_trait]
pub trait EventWatcher {
    const TAG: &'static str;
    type Middleware: providers::Middleware + 'static;
    type Contract: Deref<Target = contract::Contract<Self::Middleware>>
        + WatchableContract;
    type Events: contract::EthLogDecode;
    type Store: HistoryStore + EventHashStore;
	type EventHandler: EventHandler;

    async fn run(
        &self,
        client: Arc<Self::Middleware>,
        store: Arc<Self::Store>,
        contract: Self::Contract,
    ) -> anyhow::Result<()> {
			// Run a loop where it polls the events for each block
			// and then sends each event to the `handle_event` method on each event handler.
			// if it matches the filters.
	}
}

```

```rust
/// A trait for watching events from a watchable contract.
/// EventWatcher trait exists for deployments that are smart-contract / EVM based
#[async_trait::async_trait]
pub trait EventHandler {
	const TAG: &'static str;
    type Middleware: providers::Middleware + 'static;
    type Contract: Deref<Target = contract::Contract<Self::Middleware>>
        + WatchableContract;
    type Events: contract::EthLogDecode;
    type Store: HistoryStore + EventHashStore;

	async fn handle_event(
        &self,
        store: Arc<Self::Store>,
        contract: &Self::Contract,
        (event, log): (Self::Events, contract::LogMeta),
    ) -> anyhow::Result<()>;
}
```

Then the usage would be simply plug in the `EventHandler`s like so:

```rust
pub struct LeavesHandler;
pub struct AnchorUpdateHandler;

impl EventHandler for LeavesHandler {

	async fn handle_event(
        &self,
        store: Arc<Self::Store>,
        contract: &Self::Contract,
        (event, log): (Self::Events, contract::LogMeta),
    ) -> anyhow::Result<()> { ... }
}

impl EventHandler for AnchorUpdateHandler {
		async fn handle_event(
        &self,
        store: Arc<Self::Store>,
        contract: &Self::Contract,
        (event, log): (Self::Events, contract::LogMeta),
    ) -> anyhow::Result<()> { ... }
}

impl EventWatcher for VAnchorWatcher {
	const TAG: &'static str = "VAnchor Watcher";
    type Middleware = HttpProvider;
    type Contract = VAnchorContractWrapper<Self::Middleware>;
    type Events = VAnchorContractEvents;
    type Store = SledStore;
	type EventHandler = (LeavesHandler, AnchorUpdateHandler)
    ...
}
```

### PR DESCRIPTION
**Pull request by shekohex**
## Summary of changes
- [x] New Events Watcher API For EVM
- [ ] New Events Watcher API For Substrate

### Reference issue to close (if applicable)
- Closes #225 

-----
### Code Checklist 

- [x] Tested
- [x] Documented

### CODE REVIEW

File: src/events_watcher/mod.rs [hunk 1]
```diff
@@ -154,13 +154,20 @@ pub trait EventWatcher {
...
-                                store.clone(),
-                                &contract,
-                                (event, log.clone()),
-                            )
-                            .await;
+                        // try to handle the event with all the handlers, simultaneously.
+                        //
+                        // if any of the handlers fail, it will cancel the rest of the handlers.
+                        // and we will retry the event again.
+                        let result = futures::future::try_join_all(
```
Comment (drewstone) [inline src/events_watcher/mod.rs:161]: Let's instead process all futures and simply retry the ones that fail some constant # of times. We should not cancel the rest of the handlers when one fails.

File: src/probe.rs [hunk 2]
```diff
@@ -37,4 +37,7 @@ pub enum Kind {
     /// Relayer Leaves Store state on a specific chain/node.
     #[display(fmt = "leaves_store")]
     LeavesStore,
+    /// When the Relayer sees a new Merkle Tree insertation event on a specific chain/node.
+    #[display(fmt = "mt_insert")]
+    MerkleTreeInsertation,
```
Comment (drewstone) [inline src/probe.rs:42]:
```suggestion
    MerkleTreeInsertion,
```

Comment (drewstone) [reply src/probe.rs:42]: Please fix other references to this enum

**Review by drewstone — CHANGES_REQUESTED** [review]
Needs fixes, I'm not confident seeing how you're handling the futures of handlers that we are matching our expectations. Can you provide more info on why you think we should `try_join_all`? Why would we want to cancel futures if one fails?

File: src/probe.rs (hunk 2 above)
Comment (shekohex) [reply src/probe.rs:42]: Done.

File: src/events_watcher/mod.rs (hunk 1 above)
Comment (shekohex) [reply src/events_watcher/mod.rs]: I addressed your comments here too.

**Review by drewstone — APPROVED** [review]

File: src/events_watcher/evm/signature_bridge_watcher.rs [hunk 3]
```diff
@@ -88,6 +88,9 @@ impl<M: Middleware> super::WatchableContract
 #[derive(Copy, Clone, Debug, Default)]
 pub struct SignatureBridgeContractWatcher;
 
+#[derive(Copy, Clone, Debug, Default)]
+pub struct SignatureBridgeGovernanceOwnershipTransferredHandler;
```
Comment (salman01zp) [inline src/events_watcher/evm/signature_bridge_watcher.rs:92]: can we rename this struct to something small?

Comment (shekohex) [reply src/events_watcher/evm/signature_bridge_watcher.rs:92]: It is more like each handler name now at least should describes what events that they are handing. But nonetheless, I'm open for a better name.

Comment (salman01zp) [reply src/events_watcher/evm/signature_bridge_watcher.rs:92]: nothing comes to mind at present. Anyways we can update later.

**Review by salman01zp — APPROVED** [review]
LGTM..! Great work @shekohex

**Pull request merged**